### ISSUE TITLE
Let's host NVIDIA dependencies in our own S3

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#75703]
### 🚀 The feature, motivation and pitch

To avoid SEVs where we are unable to retrieve a dependency hosted by NVIDIA (see https://github.com/pytorch/pytorch/issues/74967), we should host copies in S3 ourselves. 

### Additional context

This was not possible before but may be possible now.

**Comment by xuzhao9** [comment]
Currently our CI is depending on three Nvidia components: Nvidia kernel driver, CUDA, and cuDNN. Are we hosting all three of them?

**Comment by atalman** [comment]
@janeyx99 @seemethere I don't think this task is very straight forward. Basically we want Nvidia to be able to update CUDA version on PyTorch.   As per this task: https://github.com/pytorch/builder/issues/1042

I think some kind of sync job that would mirror dependencies from Nvidia site to AWS S3 is good idea. So that Meta employee manual step is not necessary and everything can be performed by Nvidia engineer.

**Comment by izaitsevfb** [comment]
We had a brief discussion with @atalman, and the ideal solution seems to have a proxy (or CDN) that caches the content of the given https urls automatically.

From the user perspective, instead of the direct URL:
```
wget https://developer.download.nvidia.com/compute/redist/cudnn/v8.3.2/local_installers/11.5/cudnn-linux-x86_64-8.3.2.44_cuda11.5-archive.tar.xz
```
The requested URL would look like:
```
wget https://caching.proxy/https://developer.download.nvidia.com/compute/redist/cudnn/v8.3.2/local_installers/11.5/cudnn-linux-x86_64-8.3.2.44_cuda11.5-archive.tar.xz
```

Behind the scenes the proxy can check the availability of the original URL and do a redirect or serve a cached version instead (for security, only whitelisted URLs will be cached and served).

There are multiple ways how this can be implemented, and there are some associated costs and tradeoffs. My next step will be to do the research and compile the list of options.

**Comment by janeyx99** [comment]
@atalman Yes, I see the task was not as defined as it could be. Having a mirror sync job to mirror dependencies from the nvidia site to the AWS S3 could be an option. From previous experience, some dependencies required a user to enter a password to access, so some dependencies may still require manual updating.

Regarding the caching proxy--I remember the last time we tried a squid proxy for similar dependency caching, it was a risk due to MITM (which is kind of the point of the proxy). How will this be different?

also ccing @huydhn who has been looking into caching dependencies in general.

**Comment by huydhn** [comment]
When talking about caching, there are 2 problems here that we're trying to solve. So I will try to disentangle them a bit here:

* Reliability, when a 3rd party dependency is down. Having a proxy to sync dependencies from NVIDIA site to a more reliable location like S3 sounds reasonable. However, this might not be exactly what we need in this specific case because the failure comes from `apt-get` like `apt-get update`. IMO, we need to properly setup our own apt-get mirror for this to work:
  * Identify key repos that we absolutely depend on, NVIDIA may be?
  * Mirror the repos. I haven't done this before but https://apt-mirror.github.io/ is the tool I think
* Performance, this is probably not the focus here. But having a mirror in the same AWS region definitely helps, both in term of latency and cost.

Back to the first point, I think we need to make a judgement call to see if it's worth the effort to setup our own mirrors. I tend to say yes, but folks with more historical context might argue otherwise :)

Thoughts?

**Comment by huydhn** [comment]
Found a random [link](https://github.com/NVIDIA/nvidia-docker/issues/730) on the internet in which someone tried to mirror nvidia-docker that might be useful

**Comment by huydhn** [comment]
Using up proxy might work if we can set it up in `/etc/apt/apt.conf` or using `http_proxy`. But again, we also need to make it work with `apt-get`. 

P.S: And we also have yum repos to add into the fun. But one thing at a time I guess.

**Comment by izaitsevfb** [comment]
I'd add that we don't necessarily need to use a single approach to address all the issues. As @huydhn pointed out, we have several problematic places with third-party dependencies:
* debian repositories (apt)
* yum repositories
* conda repositories
* ad-hoc requests (wget, curl)
* ???

And there are multiple approaches to address each of them, that not necessarily work well for all or even most:
* cache deps manually or semi-automatically
* proxies
* repository mirrors
* or even [removing dependency altogether](https://github.com/pytorch/pytorch/issues/74968)

**Comment by huydhn** [comment]
Agree. Let's not try to find a silver bullet here, and just address one thing at a time (and keep in mind others). Some thoughts:

* Proxy is more powerful but comes at a cost. I would love to learn the historical lesson we have with squid proxy as @janeyx99 mentions and why it didn't work
* Mirror the repo sounds easier but has an upfront setup cost. And deb, yum, and conda probably have their own ways to achieve this. However, if this works, this could be a nice project by itself

**Comment by janeyx99** [comment]
Yes, this particular task should only attempt to resolve the nvidia part, since the other third-parties have been more reliable historically. That said, I'm glad this discussion is broader and we can give ourselves good stepping stones. 

@huydhn a proxy could be the solution, but we would have to be especially conscious of security implications. I think the last person who attempted it was @osalpekar who commented here https://github.com/pytorch/pytorch/issues/73812

**Comment by izaitsevfb** [comment]
Nvidia dependencies could be further split into:
* direct download requests from https://developer.download.nvidia.com... 
     (e.g. `wget https://developer.download.nvidia.com/...`  see  [install_cudnn.sh](https://github.com/pytorch/pytorch/blob/ca96d1d447c43484f4a9e92316d38b2864a04e6e/.circleci/docker/common/install_cudnn.sh#L7) or [install_cuda.sh](https://github.com/pytorch/builder/blob/ea37833590ac1cce994498a69cae60030571d1ff/common/install_cuda.sh#L20)) for example.
* `yum` and `apt` dependencies on https://nvidia.github.io/... 
     (e.g. [setup_ci_environment.sh](https://github.com/pytorch/pytorch/blob/c978b609f77740bb3ce7264d3d5638dd03d853bc/.circleci/scripts/setup_ci_environment.sh#L44))

I'm thinking about the following solution to address **only** the former:

* A script that can be triggered publicly by anyone, including Nvidia engineers (could be hosted on AWS lamda)
* Script takes an URL from (e.g.  https://developer.download.nvidia.com/... ) as a parameter 
* Script validates the URL agains a whitelist (e.g. regexes/masks like `https://developer.download.nvidia.com/*`)
* Script caches the given URL on our S3 if it's not already cached and returns the public S3 url
* On the client (in our build scripts) an ad-hoc fallback can be used: `(wget https://developer.download.nvidia.com/...  || wget https://s3.amazonaws.com/ossci-linux/...)`

I can see the following **advantages** to this approach:
* addressees @atalman's concern regarding our involvement in s3 dependencies caching
* somewhat better transparency compared to a proxy (addresses @janeyx99 concern re: MITM)

The **disadvantages**:
* Implementation effort
* The script will add some maintenance burden onto us (how much?)
* Clear workflow only for public dependencies (supporting dependencies under authentication may be tricky)

May I have some feedback on that? @janeyx99, @huydhn

**Comment by huydhn** [comment]
Using a glue script here is fine IMO (the whole CI flow is just a bunch of scripts working together :P). Here are some high level notes I have:

* I think we should run this script as a GitHub workflow i.e. [update_s3_htmls.yml](https://github.com/pytorch/pytorch/blob/2debdfdce63a151d8f038b8b9e8123f0b8af0820/.github/workflows/update_s3_htmls.yml). It could be daily, weekly, or even longer.
* I need to learn how our S3 bucket is setup. Having a public readable one sounds scary to me because it costs $$$. If possible, we should only allow our GitHub runners to access the link.
* After we have the confident that the flow works, I think we should just switch to use our S3 entirely for performance and simplicity.
  * S3 is much more durable than most websites including nvidia
  * It's faster and also cheaper to read from our S3 bucket

**Comment by janeyx99** [comment]
happy with the script idea as well. We do have a publicly accessible S3 heh, and we already host many other dependencies on there, such as for windows build.

**Comment by ZainRizvi** [comment]
+1 to the idea of identifying the dependencies that fail frequently and using scripts to tailor a solution to them.  

While a comprehensive proxy would be really nice (and could end up being the route we take one day in the future) it would also come with a bunch of maintenance headaches which I'd prefer to avoid for as long as possible.
